addons/l10n_in_withholding/models/account_move.py

```diff
@@ -49,6 +51,63 @@ def _compute_l10n_in_total_withholding_amount(self):
             move.l10n_in_total_withholding_amount = sum(move.l10n_in_withhold_move_ids.filtered(
                 lambda m: m.state == 'posted').l10n_in_withholding_line_ids.mapped('l10n_in_withhold_tax_amount'))
 
+    def _get_l10n_in_invalid_tax_lines(self):
+        self.ensure_one()
+        if self.country_code == 'IN' and not pan.is_valid(self.commercial_partner_id.l10n_in_pan):
+            lines = self.env['account.move.line']
+            for line in self.invoice_line_ids:
+                for tax in line.tax_ids:
+                    if (
+                        tax.l10n_in_section_id.tax_source_type == 'tcs'
+                        and tax.amount != max(tax.l10n_in_section_id.l10n_in_section_tax_ids, key=lambda t: abs(t.amount)).amount
+                    ):
+                        lines |= line._origin
+            return lines
+
+    @api.depends('invoice_line_ids.tax_ids', 'commercial_partner_id.l10n_in_pan')
+    def _compute_l10n_in_warning(self):
+        super()._compute_l10n_in_warning()
+        for move in self:
+            warnings = move.l10n_in_warning or {}
+            lines = move._get_l10n_in_invalid_tax_lines()
+            if lines:
+                warnings['lower_tcs_tax'] = {
+                        'message': _("As the Partner's PAN missing/invalid apply TCS at the higher rate."),
+                        'action_text': _("View Journal Items(s)"),
+                        'action': lines._get_records_action(
+                            name=_("Journal Items(s)"),
+                            target='current',
+                            views=[(self.env.ref("l10n_in_withholding.view_move_line_tree_l10n_in").id, "list")],
+                            domain=[('id', 'in', lines.ids)]
+                        )
+                }
+                move.l10n_in_warning = warnings
+
+    def action_l10n_in_apply_higher_tax(self):
+        self.ensure_one()
+        invalid_lines = self._get_l10n_in_invalid_tax_lines()
+        for line in invalid_lines:
+            updated_tax_ids = []
+            for tax in line.tax_ids:
+                if tax.l10n_in_section_id.tax_source_type == 'tcs':
+                    max_tax = max(
+                        tax.l10n_in_section_id.l10n_in_section_tax_ids,
+                        key=lambda t: t.amount
+                    )
+                    updated_tax_ids.append(max_tax.id)
+                else:
+                    updated_tax_ids.append(tax.id)
+            if line.tax_ids.ids != updated_tax_ids:
```

Comment:
```py
In [1]: [1,2,3] == [2,1,3]
Out[1]: False

In [2]: {1,2,3} == {2,1,3}
Out[2]: True
```
I believe comparing list to list, is an incorrect way to compare. Better to use set. Well orm itself uses set to compare the record sets too (initially was thinking to compare out the recordsets was a bad idea)

https://github.com/odoo/odoo/blame/f431a0edbf1f6fbd6d9b0f4ee2d968ce1e69dffc/odoo/models.py#L6739

Reply:
Maybe if tcs_tax := tax_ids.filtered(...): 

(Should not there be only one)
tax_ids - tcs_tax  + new_tax to set

Reply:
Directly operating on the recordsets will create a new recordsets for each operations. I guess it's better we directly set the ids as we have now. Just need to change `list->set` comparison